// @nomicfoundation/ignition-core/dist/src/internal/reconciliation/reconciler.js
import { HardhatError } from "@nomicfoundation/hardhat-errors";
import { ExecutionStatus } from "../execution/types/execution-state.js";
import { AdjacencyListConverter } from "../utils/adjacency-list-converter.js";
import { AdjacencyList } from "../utils/adjacency-list.js";
import { getFuturesFromModule } from "../utils/get-futures-from-module.js";
import { reconcileCurrentAndPreviousTypeMatch } from "./reconcile-current-and-previous-type-match.js";
import { reconcileDependencyRules } from "./reconcile-dependency-rules.js";
import { reconcileFutureSpecificReconciliations } from "./reconcile-future-specific-reconciliations.js";
export class Reconciler {
    static async reconcile(module, deploymentState, deploymentParameters, accounts, deploymentLoader, artifactResolver, defaultSender, strategy, strategyConfig) {
        const reconciliationFailures = await this._reconcileEachFutureInModule(module, {
            deploymentState,
            deploymentParameters,
            accounts,
            deploymentLoader,
            artifactResolver,
            defaultSender,
            strategy,
            strategyConfig,
        }, [
            reconcileCurrentAndPreviousTypeMatch,
            reconcileDependencyRules,
            reconcileFutureSpecificReconciliations,
        ]);
        // TODO: Reconcile sender of incomplete futures.
        const missingExecutedFutures = this._missingPreviouslyExecutedFutures(module, deploymentState);
        return { reconciliationFailures, missingExecutedFutures };
    }
    static checkForPreviousRunErrors(deploymentState) {
        const failuresOrTimeouts = Object.values(deploymentState.executionStates).filter((exState) => exState.status === ExecutionStatus.FAILED ||
            exState.status === ExecutionStatus.TIMEOUT);
        return failuresOrTimeouts.map((exState) => ({
            futureId: exState.id,
            failure: this._previousRunFailedMessageFor(exState),
        }));
    }
    static _previousRunFailedMessageFor(exState) {
        if (exState.status === ExecutionStatus.FAILED) {
            return `The previous run of the future ${exState.id} failed, and will need wiped before running again`;
        }
        if (exState.status === ExecutionStatus.TIMEOUT) {
            return `The previous run of the future ${exState.id} timed out, and will need wiped before running again`;
        }
        throw new HardhatError(HardhatError.ERRORS.IGNITION.RECONCILIATION.INVALID_EXECUTION_STATUS, {
            status: exState.status,
        });
    }
    static async _reconcileEachFutureInModule(module, context, checks) {
        // TODO: swap this out for linearization of execution state
        // once execution is fleshed out.
        const futures = this._getFuturesInReverseTopoligicalOrder(module);
        const failures = [];
        for (const future of futures) {
            const exState = context.deploymentState.executionStates[future.id];
            if (exState === undefined) {
                continue;
            }
            const result = await this._check(future, exState, context, checks);
            if (result.success) {
                continue;
            }
            failures.push(result.failure);
        }
        return failures;
    }
    static _missingPreviouslyExecutedFutures(module, deploymentState) {
        const moduleFutures = new Set(getFuturesFromModule(module).map((f) => f.id));
        const previouslyStarted = Object.values(deploymentState.executionStates).map((es) => es.id);
        const missing = previouslyStarted.filter((sf) => !moduleFutures.has(sf));
        return missing;
    }
    static _getFuturesInReverseTopoligicalOrder(module) {
        const futures = getFuturesFromModule(module);
        const adjacencyList = AdjacencyListConverter.buildAdjacencyListFromFutures(futures);
        const sortedFutureIds = AdjacencyList.topologicalSort(adjacencyList).reverse();
        return sortedFutureIds
            .map((id) => futures.find((f) => f.id === id))
            .filter((x) => x !== undefined);
    }
    static async _check(future, executionState, context, checks) {
        for (const check of checks) {
            const result = await check(future, executionState, context);
            if (result.success) {
                continue;
            }
            return result;
        }
        return { success: true };
    }
}
